casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+
+    def _get_topk_beams(
+        self, scores: torch.tensor, tokens: torch.tensor, batch: int, idx: int
+    ) -> Tuple[torch.tensor, torch.tensor]:
+        """
+        Find top-k beams with highest confidence scores and continue decoding from those.
+        Discontinue decoding for beam where stop token was predicted.
+
+        Parameters
+        ----------
+        scores : torch.Tensor of shape (n_spectra*  n_beams, max_length, n_amino_acids)
+            Output scores of the model.
+        tokens : torch.Tensor of shape (n_spectra* n_beams, max_length)
+            Output token of the model corresponding to amino acid sequences of peptides.
+        batch: int
+            Number of spectra in the batch
+        idx : int
+            Index to be considered in the current decoding step
+
+        Returns
+        -------
+        scores : torch.Tensor of shape (n_spectra, max_length, n_amino_acids, n_beams,)
+            Output scores of the model.
+        tokens : torch.Tensor of shape (n_spectra, max_length,  n_beams)
+            Output token of the model corresponding to amino acid sequences of peptides.
+        """
+        beam = self.n_beams  # S
+        vocab = self.decoder.vocab_size + 1  # V
+
+        # Reshape to group by spectrum (B for "batch")
+        scores = einops.rearrange(scores, "(B S) L V -> B L V S", S=beam)
+        tokens = einops.rearrange(tokens, "(B S) L -> B L S", S=beam)
+        prev_tokens = einops.repeat(
+            tokens[:, :idx, :],
+            "B L S -> B L V S",
+            V=vocab,
+        )
+
+        # Get the previous tokens and scores:
+        prev_scores = torch.gather(
+            scores[:, :idx, :, :],
+            dim=2,
+            index=prev_tokens,
+        )
+        prev_scores = einops.repeat(
+            prev_scores[:, :, 0, :],
+            "B L S -> B L (V S)",
+            V=vocab,
+        )
+
+        # Get scores for all possible beams at this step
+        step_scores = torch.zeros(batch, idx + 1, beam * vocab)
+        step_scores = step_scores.type_as(scores)
+        step_scores[:, :idx, :] = prev_scores
+        step_scores[:, idx, :] = einops.rearrange(
+            scores[:, idx, :, :],
+            "B V S -> B (V S)",
+        )
+
+        # Mask out terminated beams. Include delta mass induced termination
+        extended_prev_tokens = einops.repeat(
+            tokens[:, : idx + 1, :], "B L S -> B L V S", V=vocab
+        )
+        finished_mask = (
+            einops.rearrange(extended_prev_tokens, "B L V S -> B L (V S)")
+            == self.stop_token
+        ).any(axis=1)
+        # Mask out the index '0', i.e. padding token, by default
+        finished_mask[:, :beam] = True
+
+        # Figure out the top K decodings
+        _, top_idx = torch.topk(
+            step_scores.nanmean(dim=1) * (~finished_mask).int().float(),
+            beam,
+        )
+        V_idx, S_idx = np.unravel_index(top_idx.cpu(), (vocab, beam))
```

Comment:
**Issue:** Don't start variable names with a capital letter, as per the Python style convention.